File: research/python/beluga/speaker.py
"""
beluga.speaker — data models for speakers, listeners, and Beluga projects.

Spec references: §11 (Speaker), §15 (Listener), §60 (BelugaProject), §61 (config storage).
"""

from __future__ import annotations

import json
import uuid
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from .geometry import Orientation, Vector3

# Default speaker categories (spec §14)
SPEAKER_CATEGORIES = (
    "Generic",
    "Active",
    "Passive",
    "Bookshelf",
    "Floorstanding",
    "Ceiling",
    "Subwoofer",
    "Laptop",
    "Custom",
)


def _new_id() -> str:
    return str(uuid.uuid4())


@dataclass
class Room:
    """Rectangular room model (spec §6, §60).

    For 0.1 we support manual rectangular rooms only;
    imported mesh support arrives in 0.2.
    """

    name: str = ""
    length: float = 5.0  # meters, along Y
    width: float = 4.0   # meters, along X
    height: float = 2.8  # meters, along Z

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "length": self.length,
            "width": self.width,
            "height": self.height,
        }

    @staticmethod
    def from_dict(d: dict) -> "Room":
        return Room(
            name=d.get("name", ""),
            length=d.get("length", 5.0),
            width=d.get("width", 4.0),
            height=d.get("height", 2.8),
        )


@dataclass
class Speaker:
    """A single independently controllable loudspeaker (spec §11)."""

    id: str = field(default_factory=_new_id)
    name: str = "Speaker"
    category: str = "Generic"
    position: Vector3 = field(default_factory=lambda: Vector3(0.0, 0.0, 0.0))
    orientation: Orientation = field(default_factory=Orientation)
    enabled: bool = True

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "category": self.category,
            "position": self.position.to_list(),
            "orientation": self.orientation.to_dict(),
            "enabled": self.enabled,
        }

    @staticmethod
    def from_dict(d: dict) -> "Speaker":
        pos = d.get("position", [0.0, 0.0, 0.0])
        return Speaker(
            id=d.get("id", _new_id()),
            name=d.get("name", "Speaker"),
            category=d.get("category", "Generic"),
            position=Vector3(pos[0], pos[1], pos[2]),
            orientation=Orientation.from_dict(d.get("orientation", {})),
            enabled=d.get("enabled", True),
        )


@dataclass
class Listener:
    """A listener position and orientation (spec §15-§17).

    The acoustic origin is approximately the center point between the ears,
    placed at `position + (0, 0, ear_height)` above `position` if the
    `position` tag designates a chair/couch seat level. For 0.1 simplicity,
    `position` already denotes the ear-center position in room coordinates.
    `ear_height` is stored for future refinement and UX.
    """

    id: str = field(default_factory=_new_id)
    name: str = "Listener"
    position: Vector3 = field(default_factory=lambda: Vector3(0.0, 0.0, 1.1))
    orientation: Orientation = field(default_factory=Orientation)
    ear_height: float = 1.10  # meters

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "name": self.name,
            "position": self.position.to_list(),
            "orientation": self.orientation.to_dict(),
            "ear_height": self.ear_height,
        }

    @staticmethod
    def from_dict(d: dict) -> "Listener":
        pos = d.get("position", [0.0, 0.0, 1.1])
        return Listener(
            id=d.get("id", _new_id()),
            name=d.get("name", "Listener"),
            position=Vector3(pos[0], pos[1], pos[2]),
            orientation=Orientation.from_dict(d.get("orientation", {})),
            ear_height=d.get("ear_height", 1.10),
        )


@dataclass
class BelugaProject:
    """A Beluga project bundle (spec §60-§62).

    Stores room, speakers, listener, and (future) calibration/playback settings.
    Can be saved/loaded as JSON for round-trip persistence.
    """

    name: str = "Untitled"
    room: Room = field(default_factory=Room)
    speakers: list[Speaker] = field(default_factory=list)
    listeners: list[Listener] = field(default_factory=list)
    active_listener_id: str | None = None

    def active_listener(self) -> Listener | None:
        if not self.listeners:
            return None
        if self.active_listener_id:
            for l in self.listeners:
                if l.id == self.active_listener_id:
                    return l
        return self.listeners[0]

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "room": self.room.to_dict(),
            "speakers": [s.to_dict() for s in self.speakers],
            "listeners": [l.to_dict() for l in self.listeners],
            "active_listener_id": self.active_listener_id,
        }

    @staticmethod
    def from_dict(d: dict) -> "BelugaProject":
        return BelugaProject(
            name=d.get("name", "Untitled"),
            room=Room.from_dict(d.get("room", {})),
            speakers=[Speaker.from_dict(s) for s in d.get("speakers", [])],
            listeners=[Listener.from_dict(l) for l in d.get("listeners", [])],
            active_listener_id=d.get("active_listener_id"),
        )

    def save(self, path: str | Path) -> None:
        Path(path).write_text(json.dumps(self.to_dict(), indent=2), encoding="utf-8")

    @staticmethod
    def load(path: str | Path) -> "BelugaProject":
        return BelugaProject.from_dict(json.loads(Path(path).read_text(encoding="utf-8")))

File: crates/beluga-core/src/speaker.rs
//! beluga-core::speaker — Data models for speakers, listeners, rooms, projects.
//! Port of `research/python/beluga/speaker.py`.
//! Spec references: §6 (Room), §11 (Speaker), §14 (categories), §15 (Listener),
//! §17 (ear height), §60 (BelugaProject), §61 (config storage).

use serde::{Deserialize, Serialize};

use crate::geometry::{Orientation, Vector3};

/// Default speaker categories (spec §14).
pub const SPEAKER_CATEGORIES: &[&str] = &[
    "Generic",
    "Active",
    "Passive",
    "Bookshelf",
    "Floorstanding",
    "Ceiling",
    "Subwoofer",
    "Laptop",
    "Monitor",
    "Custom",
];

/// Rectangular room model (spec §6, §60).
///
/// For 0.1 we support manual rectangular rooms only;
/// imported mesh support arrives in 0.2+.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Room {
    pub name: String,
    pub length: f64, // meters, along Y
    pub width: f64,  // meters, along X
    pub height: f64, // meters, along Z
}

impl Default for Room {
    fn default() -> Self {
        Room {
            name: String::new(),
            length: 5.0,
            width: 4.0,
            height: 2.8,
        }
    }
}

impl Room {
    pub fn new(name: &str, length: f64, width: f64, height: f64) -> Self {
        Room {
            name: name.to_string(),
            length,
            width,
            height,
        }
    }
}

/// A single independently controllable loudspeaker (spec §11).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Speaker {
    pub id: String,
    pub name: String,
    pub category: String,         // "Generic", "Bookshelf", "Floorstanding", ...
    pub position: Vector3,        // room-space meters
    pub orientation: Orientation, // yaw/pitch/roll degrees
    pub enabled: bool,
}

impl Speaker {
    pub fn new(id: &str, name: &str, category: &str, position: Vector3) -> Self {
        Speaker {
            id: id.to_string(),
            name: name.to_string(),
            category: category.to_string(),
            position,
            orientation: Orientation::default(),
            enabled: true,
        }
    }
}

/// A listener position and orientation (spec §15–§17).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Listener {
    pub id: String,
    pub name: String,
    pub position: Vector3,
    pub orientation: Orientation,
    pub ear_height: f64, // meters
}

impl Listener {
    pub fn new(id: &str, name: &str, position: Vector3) -> Self {
        Listener {
            id: id.to_string(),
            name: name.to_string(),
            position,
            orientation: Orientation::default(),
            ear_height: 1.10,
        }
    }
}

/// A Beluga project bundle (spec §60-§62).
///
/// Stores room, speakers, listeners, and active listener.
/// JSON format matches the Python to_dict/from_dict:
///   position as [x, y, z], orientation as {yaw, pitch, roll}.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct BelugaProject {
    pub name: String,
    pub room: Room,
    pub speakers: Vec<Speaker>,
    pub listeners: Vec<Listener>,
    pub active_listener_id: Option<String>,
}

impl BelugaProject {
    pub fn new(name: &str) -> Self {
        BelugaProject {
            name: name.to_string(),
            room: Room::default(),
            speakers: Vec::new(),
            listeners: Vec::new(),
            active_listener_id: None,
        }
    }

    /// Returns the active listener, or the first one if none is active.
    pub fn active_listener(&self) -> Option<&Listener> {
        if self.listeners.is_empty() {
            return None;
        }
        if let Some(id) = &self.active_listener_id {
            for l in &self.listeners {
                if &l.id == id {
                    return Some(l);
                }
            }
        }
        self.listeners.first()
    }

    /// Returns the active listener index into `self.listeners`.
    pub fn active_listener_index(&self) -> Option<usize> {
        if self.listeners.is_empty() {
            return None;
        }
        if let Some(id) = &self.active_listener_id {
            for (i, l) in self.listeners.iter().enumerate() {
                if &l.id == id {
                    return Some(i);
                }
            }
        }
        Some(0)
    }

    pub fn to_json(&self) -> String {
        serde_json::to_string_pretty(self).expect("project serializes to JSON")
    }

    pub fn from_json(s: &str) -> Result<Self, String> {
        serde_json::from_str(s).map_err(|e| e.to_string())
    }
}

impl Default for BelugaProject {
    fn default() -> Self {
        BelugaProject::new("Untitled")
    }
}

// ---------------------------------------------------------------------------
// Tests — mirror research/python/tests/test_speaker.py
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn room_default() {
        let r = Room::default();
        assert_eq!(r.length, 5.0);
        assert_eq!(r.width, 4.0);
        assert_eq!(r.height, 2.8);
    }

    #[test]
    fn json_roundtrip() {
        let mut proj = BelugaProject::new("Test Room");
        proj.speakers.push(Speaker::new(
            "spk-1",
            "Left",
            "Bookshelf",
            Vector3::new(-1.0, 2.0, 0.0),
        ));
        proj.speakers.push(Speaker::new(
            "spk-2",
            "Right",
            "Bookshelf",
            Vector3::new(1.0, 2.0, 0.0),
        ));
        let listener = Listener::new("list-1", "Main", Vector3::new(0.0, 0.0, 1.1));
        proj.listeners.push(listener.clone());
        proj.active_listener_id = Some("list-1".to_string());

        let json = proj.to_json();
        let restored = BelugaProject::from_json(&json).unwrap();
        assert_eq!(restored.name, "Test Room");
        assert_eq!(restored.speakers.len(), 2);
        assert_eq!(restored.speakers[0].name, "Left");
        assert!((restored.speakers[0].position.x - (-1.0)).abs() < 1e-9);
        assert_eq!(restored.listeners.len(), 1);
        assert_eq!(restored.active_listener_id, Some("list-1".to_string()));
        assert!((restored.active_listener().unwrap().position.z - 1.1).abs() < 1e-9);
    }

    #[test]
    fn active_listener_first_when_no_id() {
        let mut proj = BelugaProject::new("Test");
        proj.listeners
            .push(Listener::new("l1", "A", Vector3::new(0.0, 0.0, 1.0)));
        proj.listeners
            .push(Listener::new("l2", "B", Vector3::new(0.0, 0.0, 2.0)));
        assert_eq!(proj.active_listener().unwrap().id, "l1");
    }

    #[test]
    fn active_listener_by_id() {
        let mut proj = BelugaProject::new("Test");
        proj.listeners
            .push(Listener::new("l1", "A", Vector3::new(0.0, 0.0, 1.0)));
        proj.listeners
            .push(Listener::new("l2", "B", Vector3::new(0.0, 0.0, 2.0)));
        proj.active_listener_id = Some("l2".to_string());
        assert_eq!(proj.active_listener().unwrap().id, "l2");
    }

    #[test]
    fn speaker_default_orientation() {
        let s = Speaker::new("s1", "Test", "Generic", Vector3::new(0.0, 0.0, 0.0));
        assert_eq!(s.orientation, Orientation::default());
        assert!(s.enabled);
    }
}
